File: crates/verify/ess-conformance/tests/fixtures/quoted-predicates.go
package essconform

import (
	"encoding/json"
	"strings"
	"testing"
)

func quotedSuite(t *testing.T, expression any) string {
	t.Helper()
	var document map[string]any
	if err := json.Unmarshal([]byte(suiteJSON), &document); err != nil {
		t.Fatal(err)
	}
	for _, raw := range document["scenarios"].(map[string]any) {
		step := raw.(map[string]any)["steps"].([]any)[0].(map[string]any)
		step["expectation"].(map[string]any)["predicate"] = expression
	}
	encoded, err := json.Marshal(document)
	if err != nil {
		t.Fatal(err)
	}
	return string(encoded)
}

func TestQuotedTrailingTokensRefused(t *testing.T) {
	for _, expression := range []string{
		`to == "" or text == ""`, `to == '' and text == ''`,
		`to == "done" trailing`, `to == "done""next"`,
		`not to == "" or text == ""`, `to == "unterminated`,
		`to == 'unterminated`, `to == "escaped\"`, `to == "slash\\" or text == ""`,
	} {
		t.Run(expression, func(t *testing.T) {
			_, err := admitRunInput(quotedSuite(t, expression))
			if err == nil || !strings.Contains(err.Error(), "structured") {
				t.Errorf("original suite admitted or lacks guidance: %v", err)
			}
			if _, err := parseLeaf(expression); err == nil {
				t.Error("evaluator admitted malformed leaf")
			}
		})
	}
}

func TestQuotedValuesRemainExact(t *testing.T) {
	for _, vector := range []struct{ expression, value string }{
		{`to == ""`, ""}, {`to == 'true'`, "true"}, {`to == "or and not"`, "or and not"},
		{" \tto == '  spaced  ' \n", "  spaced  "},
		{`to == "say \"or\""`, `say \"or\"`}, {`to == 'it\'s valid'`, `it\'s valid`},
		{`to == "path\\"`, `path\\`}, {`to == "雪 and ☃"`, "雪 and ☃"},
		{`to == "a == b or c != d"`, "a == b or c != d"},
	} {
		t.Run(vector.expression, func(t *testing.T) {
			if _, err := admitRunInput(quotedSuite(t, vector.expression)); err != nil {
				t.Fatal(err)
			}
			parsed, err := parseLeaf(vector.expression)
			if err != nil {
				t.Fatal(err)
			}
			if parsed.right.literal != vector.value {
				t.Fatalf("literal=%#v want %#v", parsed.right.literal, vector.value)
			}
			if parsed.evaluate(factSource{"to": vector.value}) != truthTrue {
				t.Fatal("exact value does not match")
			}
		})
	}
}

func TestQuotedStructuredAndPrefixNot(t *testing.T) {
	for _, expression := range []any{
		map[string]any{"any": []any{`to == ""`, map[string]any{"all": []any{`text == ''`, map[string]any{"not": "disabled"}}}}},
		`not to == ""`,
	} {
		if _, err := admitRunInput(quotedSuite(t, expression)); err != nil {
			t.Fatal(err)
		}
		parsed, err := fromNode(expression)
		if err != nil {
			t.Fatal(err)
		}
		if parsed.evaluate(factSource{"to": "filled", "text": "", "disabled": false}) != truthTrue {
			t.Fatalf("valid composition failed: %#v", expression)
		}
	}
}

func TestQuotedComparisonOperatorsInsideLiteralRemainData(t *testing.T) {
	for _, expression := range []string{`to < "a == b"`, `to < "a != b"`, `to < 'a >= b'`} {
		if _, err := admitRunInput(quotedSuite(t, expression)); err != nil {
			t.Fatal(err)
		}
		parsed, err := parseLeaf(expression)
		if err != nil {
			t.Fatal(err)
		}
		if parsed.left.path != "to" || parsed.op != "<" || parsed.evaluate(factSource{"to": "0"}) != truthTrue {
			t.Fatalf("comparison inside literal became an operator: %#v", parsed)
		}
	}
}

File: crates/verify/ess-conformance/tests/fixtures/quoted-predicates.mjs
import assert from 'node:assert/strict'
import test from 'node:test'
import {admitSuite} from '../../assets/coverage-admission.js'

const scenario = 'quoted.core/authored/operand'
const document = predicate => JSON.stringify({
  provenance: {suite_version: 'ess-conformance/5', system: 'quoted', specification_version: 'v1', spec_digest: 'a'.repeat(64), contract_digest: 'b'.repeat(64)},
  scenarios: {[scenario]: {purpose: 'Check a quoted operand', steps: [{step: 'expect_view', view: 'quoted.core.Rows', expectation: {expect: 'satisfies', predicate}}], source: []}},
  coverage: {
    selection: {scope: {kind: 'system'}, origins: 'authored', filter: {kind: 'all'}},
    knowledge: 'complete_inventory', generated: [], authored: [scenario], outside: [], refused: [],
    authored_sources: {'operand.yaml': {digest: `sha256:${'c'.repeat(64)}`, scenario, disposition: 'accepted'}},
    counts: {generated: 0, authored: 1, outside: 0, refused: 0},
  },
})
const meaning = admitted => admitted.meaning[scenario].steps[0].expectation.predicate

for (const expression of [
  'to == "" or text == ""', "to == '' and text == ''", 'to == "done" trailing',
  'to == "done""next"', 'not to == "" or text == ""', 'to == "unterminated',
  "to == 'unterminated", String.raw`to == "escaped\"`, String.raw`to == "slash\\" or text == ""`,
]) {
  test(`refuse ${expression}`, async () => {
    await assert.rejects(admitSuite(document(expression)), /structured any\/all\/not/)
  })
}

for (const [expression, literal] of [
  ['to == ""', ''], ["to == 'true'", 'true'], ['to == "or and not"', 'or and not'],
  [" \tto == '  spaced  ' \n", '  spaced  '],
  [String.raw`to == "say \"or\""`, String.raw`say \"or\"`],
  [String.raw`to == 'it\'s valid'`, String.raw`it\'s valid`],
  [String.raw`to == "path\\"`, String.raw`path\\`],
  ['to == "雪 and ☃"', '雪 and ☃'], ['to == "a == b or c != d"', 'a == b or c != d'],
]) {
  test(`preserve ${expression}`, async () => {
    assert.deepEqual(meaning(await admitSuite(document(expression))), ['compare', 'to', '==', ['literal', literal]])
  })
}

test('structured composition and prefix not remain admitted', async () => {
  await admitSuite(document({any: ['to == ""', {all: ["text == ''", {not: 'disabled'}]}]}))
  assert.deepEqual(meaning(await admitSuite(document('not to == ""'))), ['not', ['compare', 'to', '==', ['literal', '']]])
})

for (const literal of ['a == b', 'a != b', 'a >= b']) {
  test(`operator in literal ${literal}`, async () => {
    assert.deepEqual(meaning(await admitSuite(document(`to < "${literal}"`))), ['compare', 'to', '<', ['literal', literal]])
  })
}
